File: Day13/6_Next_Greater_Element.java
import java.util.Stack;
import java.util.HashMap;

class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        Map<Integer, Integer> nextGreaterMap = new HashMap<>();
        Stack<Integer> stack = new Stack<>();

        // Store next greater elements in a map
        for (int num : nums2) {
            while (!stack.isEmpty() && stack.peek() < num) {
                nextGreaterMap.put(stack.pop(), num);
            }
            stack.push(num);
        }

        // Fill result array
        int[] res = new int[nums1.length];
        for (int i = 0; i < nums1.length; i++) {
            res[i] = nextGreaterMap.getOrDefault(nums1[i], -1);
        }

        return res;
    }
}

File: Day13/6_Next_Greater_Element.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        dict1 = {}
        stack = []

        for i in nums2:
            while stack and stack[-1] < i:
                dict1[stack.pop()] = i
            stack.append(i)

        res = []
        for i in range(len(nums1)):
            if nums1[i] in dict1:
                res.append(dict1[nums1[i]])
            else:
                res.append(-1)

        return res
